README.rst
########################
djangocms-casper-helpers
########################
.. image:: https://travis-ci.org/divio/djangocms-casper-helpers.svg?branch=master
    :target: https://travis-ci.org/divio/djangocms-casper-helpers

Set of helpers for writing integration tests for django CMS applications
with `CasperJS <http://casperjs.org/>`_.


CHANGELOG.rst
==== 4.0.0 (2017-09-03) ====

- Fixed things to work with CMS 4.0


==== 3.0.0 (2017-11-15) ====

- Changed `addPage` / `removePage` in backwards incompatible way due to page
  tree refactor on the cms


==== 2.0.0 (2017-09-21) ====

- deprecated casper-summoner
- fixed path resolution for casper binary
- added CMS 3.5 helpers


==== 1.4.0 (2017-06-13) ====

- Allowed sleep timeout to be passed as an option
- Increased default sleep timeout to 2 minutes


==== 1.3.0 (2016-08-24) ====

- Updated helpers to work with CMS 3.4
- Added ``waitUntilActionsDropdownLoaded`` method


==== 1.2.0 (2016-07-14) ====

- Server now looks for a free port before starting to avoid clashes
- Exposed current port via settings


==== 1.1.2 (2016-06-09) ====

- Changed ``createJSTreeXPathFromTree`` to start from the root of the tree.


==== 1.1.1 (2016-06-03) ====

- Use namespaced CMS jQuery


==== 1.1.0 (2016-05-25) ====

- Added "getXPathForAdminSection"


==== 1.0.4 (2016-05-24) ====

- Revert casperjs version to 1.1.0-beta5


==== 1.0.3 (2016-05-22) ====

- Reorganized file structure and moved to a separate repo
- Added linting


==== 1.0.2 (2016-05-22) ====

- Fixed a bug with multipl Phantomjs installations


==== 1.0.1 (2016-05-22) ====

- Added possibility to specify own paths to Casper and Phantom


==== 1.0.0 (2016-05-22) ====

- Initial release


settings.js
'use strict';

// #############################################################################
// Global values, used throughout the test suites
var DEFAULT_PORT = 8000;

if (casper.cli.has('server-port')) {
    DEFAULT_PORT = casper.cli.get('server-port');
}

/**
 * @function configure
 * @param {Object} [opts] options
 * @param {Number|String} [opts.port] port where server runs
 * @param {String} [opts.editOn] edit mode "on" get param
 * @param {String} [opts.editOff] options edit mode "off" get param
 * @returns {Object} config
 */
function configure(opts) {
    var port = opts && typeof opts.port !== 'undefined' ? opts.port : DEFAULT_PORT;
    var editOn = opts && typeof opts.editOn !== 'undefined' ? opts.editOn : 'edit';
    var editOff = opts && typeof opts.editOff !== 'undefined' ? opts.editOff : 'edit_off';
    var config = {
        adminTitle: 'Log in | Django site admin',
        // TODO:
        // - configure languages with djangocms-helper
        // - remove hardcoded host, port, language values
        baseUrl: 'http://localhost:' + port + '/en/',
        editUrl: 'http://localhost:' + port + '/en/?' + editOn,
        editOffUrl: 'http://localhost:' + port + '/en/?' + editOff,
        adminUrl: 'http://localhost:' + port + '/en/admin/login/',
        adminLogoutUrl: 'http://localhost:' + port + '/en/admin/logout/',
        adminPagesUrl: 'http://localhost:' + port + '/en/admin/cms/pagecontent/',
        adminPageUrl: 'http://localhost:' + port + '/en/admin/cms/page/',
        adminUsersUrl: 'http://localhost:' + port + '/en/admin/auth/user/',
        port: port,
        credentials: {
            username: 'admin',
            password: 'admin'
        },
        content: {
            page: {
                title: 'First page',
                text: 'First page content'
            }
        },
        user: {
            firstName: 'test-first-name',
            lastName: 'test-last-name',
            userEmail: 'test@email.com',
            username: 'test-add-user',
            password: 'test'
        }
    };

    return config;
}

module.exports = configure();
module.exports.configure = configure;


.eslintrc.js
module.exports = {
    "env": {
        "browser": true,
        "node": true,
        "jquery": true,
        "jasmine": true
    },
    "globals": {
        "CMS": true,
        "casper": true
    },
    "root": true,
    "ecmaFeatures": {
        "modules": true
    },
    "rules": {
        // Possible Errors
        "comma-dangle": [2, "never"],
        "no-cond-assign": 2,
        "no-console": 1,
        "no-constant-condition": 2,
        "no-control-regex": 2,
        "no-debugger": 2,
        "no-dupe-args": 2,
        "no-dupe-keys": 2,
        "no-duplicate-case": 2,
        "no-empty-character-class": 2,
        "no-empty": ["error", { "allowEmptyCatch": true }],
        "no-ex-assign": 2,
        "no-extra-boolean-cast": 2,
        "no-extra-parens": ["error", "all", {
            "nestedBinaryExpressions": false
        }],
        "no-extra-semi": 2,
        "no-func-assign": 2,
        "no-inner-declarations": 2,
        "no-invalid-regexp": 2,
        "no-irregular-whitespace": 2,
        "no-negated-in-lhs": 2,
        "no-obj-calls": 2,
        "no-regex-spaces": 2,
        "no-sparse-arrays": 2,
        "no-unexpected-multiline": 2,
        "no-unreachable": 2,
        "use-isnan": 2,
        "valid-jsdoc": [2, {
            "requireReturn": false,
            "requireParamDescription": false,
            "requireReturnDescription": false,
            "prefer": {
                "return": "returns"
            }
        }],
        "valid-typeof": 2,

        // Best Practices
        "accessor-pairs": 2,
        "block-scoped-var": 2,
        "complexity": ["error", { "max": 10 } ],
        "consistent-return": 0,
        "curly": 2,
        "default-case": 2,
        "dot-location": [2, "property"],
        "dot-notation": 2,
        "eqeqeq": 2,
        "guard-for-in": 2,
        "no-alert": 2,
        "no-caller": 2,
        "no-case-declarations": 2,
        "no-div-regex": 2,
        "no-else-return": 1,
        "no-empty-pattern": 2,
        "no-eq-null": 2,
        "no-eval": 2,
        "no-extend-native": 2,
        "no-extra-bind": 2,
        "no-fallthrough": 2,
        "no-floating-decimal": 2,
        "no-implicit-coercion": 0,
        "no-implied-eval": 2,
        "no-invalid-this": 0,
        "no-iterator": 2,
        "no-labels": 2,
        "no-lone-blocks": 2,
        "no-loop-func": 2,
        "no-magic-numbers": 0,
        "no-multi-spaces": 2,
        "no-multi-str": 0,
        "no-native-reassign": 2,
        "no-new-func": 2,
        "no-new-wrappers": 2,
        "no-new": 0,
        "no-octal-escape": 2,
        "no-octal": 2,
        "no-param-reassign": 2,
        "no-process-env": 0,
        "no-proto": 2,
        "no-redeclare": 2,
        "no-return-assign": 2,
        "no-script-url": 2,
        "no-self-compare": 2,
        "no-sequences": 2,
        "no-throw-literal": 2,
        "no-unused-expressions": [2, { "allowShortCircuit": true }],
        "no-useless-call": 2,
        "no-useless-concat": 2,
        "no-void": 2,
        "no-warning-comments": 0,
        "no-with": 2,
        "radix": 2,
        "vars-on-top": 0,
        "wrap-iife": [2, "inside"],
        "yoda": [2, "never", { "exceptRange": true }],

        // Strict Mode
        "strict": [2, "global"],

        // Variables
        "init-declarations": 0,
        "no-catch-shadow": 2,
        "no-delete-var": 2,
        "no-label-var": 2,
        "no-shadow-restricted-names": 2,
        "no-shadow": 2,
        "no-undef-init": 2,
        "no-undef": 2,
        "no-undefined": 0,
        "no-unused-vars": 2,
        "no-use-before-define": 2,

        // Stylistic Issues
        "array-bracket-spacing": [2, "never"],
        "block-spacing": 2,
        "brace-style": [2, "1tbs"],
        "camelcase": 0,
        "comma-spacing": [2, {"before": false, "after": true}],
        "comma-style": [2, "last"],
        "computed-property-spacing": [2, "never"],
        "consistent-this": [2, "that"],
        "eol-last": 2,
        "func-names": 0,
        "func-style": 0,
        "id-length": 0,
        "id-match": 0,
        "indent": ["error", 4, {
            "SwitchCase": 1
        }],
        "jsx-quotes": 0,
        "key-spacing": [2, {"beforeColon": false, "afterColon": true}],
        "linebreak-style": [2, "unix"],
        "lines-around-comment": 0,
        "max-nested-callbacks": [2, 8],
        "new-cap": 2,
        "new-parens": 2,
        "newline-after-var": 2,
        "no-array-constructor": 2,
        "no-continue": 2,
        "no-inline-comments": 0,
        "no-lonely-if": 2,
        "no-mixed-spaces-and-tabs": 2,
        "no-multiple-empty-lines": [2, {"max": 2}],
        "no-negated-condition": 2,
        "no-nested-ternary": 2,
        "no-new-object": 2,
        "no-restricted-syntax": 0,
        "no-spaced-func": 0,
        "no-ternary": 0,
        "no-trailing-spaces": 2,
        "no-underscore-dangle": 0,
        "no-unneeded-ternary": 2,
        "object-curly-spacing": [2, "always", {
            "objectsInObjects": true,
            "arraysInObjects": true
        }],
        "one-var": [2, "never"],
        "operator-assignment": 2,
        "operator-linebreak": [2, "after"],
        "padded-blocks": 0,
        "quote-props": [2, "consistent-as-needed"],
        "quotes": [2, "single", "avoid-escape"],
        "require-jsdoc": 2,
        "semi-spacing": [2, {"before": false, "after": true}],
        "semi": [2, "always"],
        "sort-vars": 0,
        "keyword-spacing": 2,
        "space-before-blocks": 2,
        "space-before-function-paren": ["error", { "anonymous": "always", "named": "never" }],
        "space-in-parens": [2, "never"],
        "space-infix-ops": 2,
        "space-unary-ops": 2,
        "spaced-comment": 2,
        "wrap-regex": 2,

        // ES6
        "arrow-parens": [2, "always"],

        // Legacy
        "max-depth": [2, 4],
        "max-len": [2, 120],
        "max-params": [2, 3],
        "max-statements": 0,
        "no-bitwise": 2,
        "no-plusplus": 0
    }
}


random-string.js
'use strict';

/**
 * Generates random strings
 *
 * @param {Object} [options] options
 * @param {Number} [options.length=6] length of the string to produce
 * @param {Boolean} [options.onlyDigits=false] make helper return string containing numbers only
 * @param {Boolean} [options.withWhitespaces=false] make helper return string containing numbers only
 * @returns {String} randomString
 */
module.exports.randomString = function (options) {
    // defaults
    var opts = options || {};
    var stringLength = opts.length || 6;
    var onlyDigits = opts.onlyDigits || false;
    var withWhitespaces = opts.withWhitespaces || false;
    var randomString = '';
    var possibleCharacters = '0123456789';
    var i;

    if (stringLength <= 0) {
        return '';
    }
    if (!onlyDigits) {
        possibleCharacters += 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
    }
    if (withWhitespaces) {
        possibleCharacters += ' ';
    }
    for (i = 0; i < stringLength; i++) {
        randomString += possibleCharacters.charAt(Math.floor(Math.random() * possibleCharacters.length));
    }

    return randomString;
};


gulp/index.js
'use strict';

var terminate = require('terminate');
var Promise = require('bluebird');
var _ = require('lodash');
var path = require('path');
var child_process = require('child_process');
var spawn = require('child_process').spawn;
var portfinder = require('portfinder');

module.exports = function (options) {
    var INTEGRATION_TESTS = options.tests;
    var pathToBin = child_process.execSync('npm bin').toString().trim();
    var argv = options.argv || {};
    var logger = options.logger;
    var pathToPhantom = options.pathToPhantom || require('phantomjs-prebuilt').path;
    var pathToCasper = options.pathToCasper || path.join(pathToBin, 'casperjs');
    var waitForMigrations = options.waitForMigrations || 120;
    var serverPort = 8000;

    var integrationTests = {
        /**
        * Runs pyhon testserver.py and sleeps for a minute to let it run migrations.
        * Respects `clean` cli argument to remove the existing local database.
        *
        * @method startServer
        * @param {String} args plain string of arguments to be passed to testserver.py (space separated)
        * @returns {Promise} fullfilled when sleep ends
        */
        startServer: function startServer(args) {
            return new Promise(function (resolve) {
                if (argv && argv.clean) {
                    child_process.execSync('rm -rf ' + options.dbPath);
                }
                if (argv && argv.server === false) {
                    resolve(false);
                    return;
                }

                portfinder.basePort = serverPort;

                portfinder.getPort(function (err, port) {
                    var server = spawn(
                        'python',
                        options.serverCommand
                            .split(' ')
                            .concat(args.split(' '))
                            .concat(['--port=' + port]));

                    logger('\n\nStarting a server on http://localhost:' + port + '\n\n');

                    serverPort = port;

                    server.stdout.on('data', function (data) {
                        // eslint-disable-next-line
                        console.log(data.toString().slice(0, -1));
                    });

                    server.stderr.on('data', function (data) {
                        logger('Server: ', data.toString().slice(0, -1));
                    });

                    var sleep = spawn('sleep', [waitForMigrations]);

                    sleep.on('close', function () {
                        resolve(server.pid);
                    });
                });
            });
        },

        /**
        * Prepares files for consumption by casper. Input given
        * as array of buckets, buckets being array of strings (file names) or
        * objects (filename + custom server args). Converts strings to objects
        * with { file: originalString, serverArgs: '' }
        *
        * @method prepareBuckets
        * @returns {Array<Array<Object>>} modified buckets
        * @example
        * input:
        * [
        *     [ 'file1', 'file2' ],
        *     [ 'file3', { file: 'file4', serverArgs: '--something' } ]
        * ]
        *
        * output:
        * [
        *     [ { file: 'file1', serverArgs: '' }, { file: 'file2', serverArgs: '' } ],
        *     [ { file: 'file3', serverArgs: '' }, { file: 'file4', serverArgs: '--something' } ]
        * ]
        */
        prepareBuckets: function prepareBuckets() {
            return INTEGRATION_TESTS.map(function (bucket) {
                return bucket.map(function (test) {
                    if (typeof test === 'object') {
                        return test;
                    }

                    return {
                        file: test,
                        serverArgs: ''
                    };
                });
            });
        },

        /**
        * Prepares files for testing. Respects INTEGRATION_TESTS_BUCKET env
        * variable (on travis we run tests in separate jobs to speed up the whole suite)
        * and cli arguments. If there are multiple tests that require same server (same server arguments)
        * it would group them together for speed.
        *
        * @method prepareFiles
        * @returns {Promise} immediately resolved with prepared files object
        * @example {
        *     "": [ { file: 'path1', serverArgs: "" } ],
        *     "--something": [ { file: 'path2', serverArgs: "--something" } ]
        * }
        */
        prepareFiles: function prepareFiles() {
            var buckets = this.prepareBuckets();

            var files = [];

            // on travis we split up integration tests into three buckets,
            // and set which bucket will be used through environment variable
            switch (process.env.INTEGRATION_TESTS_BUCKET) {
                case '1':
                case '2':
                case '3':
                    files = buckets[Number(process.env.INTEGRATION_TESTS_BUCKET) - 1];
                    break;
                default:
                    files = buckets.reduce(function (memo, bucket) {
                        return memo.concat(bucket);
                    }, []);
            }

            var pre = [{
                file: options.pathToTests + '/integration/setup.js'
            }];

            var fileNames;

            if (argv && argv.tests) {
                fileNames = argv.tests.split(',');
                logger('Running tests for ' + fileNames.join(', '));
                files = fileNames.map(function (fileName) {
                    return _.find(files, function (file) {
                        return file.file === fileName;
                    }) || {
                        file: fileName,
                        serverArgs: ''
                    };
                });
            }

            var tests = files.map(function (file) {
                return _.merge({}, file, {
                    file: options.pathToTests + '/integration/' + file.file + '.js'
                });
            });

            var groupedTests = _.mapValues(_.groupBy(tests, 'serverArgs'), function (testsArray) {
                return pre.concat(testsArray);
            });

            return Promise.resolve(groupedTests);
        },

        /**
        * Runs casperjs process with tests passed as arguments to it and logs output.
        *
        * @method runTests
        * @param {String[]} tests paths to tests
        * @returns {Promise} resolves with casper exit code (0 or 1)
        */
        runTests: function (tests) {
            return new Promise(function (resolve) {
                var visual = argv && argv.visual;
                var visualPort = argv && argv.visualPort || '8002';
                var params = ['test', '--web-security=no', '--server-port=' + serverPort];

                if (visual) {
                    params.push('--visual=' + (visual || 10));
                    params.push('--visual-port=' + visualPort);
                }
                var casperChild = spawn(
                    pathToCasper,
                    params.concat(tests)
                );

                casperChild.stdout.on('data', function (data) {
                    logger('CasperJS:', data.toString().slice(0, -1));
                });

                casperChild.on('close', function (code) {
                    resolve(code);
                });
            });
        }
    };

    return function (done) {
        process.env.PHANTOMJS_EXECUTABLE = pathToPhantom;

        integrationTests
            .prepareFiles()
            .then(function (groupedTests) {
                return Promise.reduce(Object.keys(groupedTests), function (items, serverArgs) {
                    var tests = groupedTests[serverArgs].map(function (obj) {
                        return obj.file;
                    });
                    var serverPid;

                    return integrationTests.startServer(serverArgs)
                        .then(function (pid) {
                            serverPid = pid;
                        })
                        .then(function () {
                            return integrationTests.runTests(tests);
                        })
                        .then(function (exitCode) {
                            return new Promise(function (resolve, reject) {
                                var finish = function () {
                                    if (exitCode === 0) {
                                        resolve(items);
                                    } else {
                                        reject('Failure');
                                    }
                                };

                                if (serverPid) {
                                    terminate(serverPid, finish);
                                } else {
                                    finish();
                                }
                            });
                        });
                }, []);
            })
            .then(function () {
                done(0);
            })
            .catch(function (e) {
                // eslint-disable-next-line
                logger(e);
                done(1);
            });
    };
};
